lib/dry/configurable/setting.rb

```diff
@@ -94,6 +96,12 @@ def pristine
         with(input: Undefined)
       end
 
+      # @api private
+      def finalize!
+        value.is_a?(Config) ? value.finalize! : value.freeze
```

Comment:
I don't think that freezing values is the right way. You can't know in advance what values are, it can be anything. For instance, if it's a module or a class (a reasonable assumption for config!) then you won't be able to define new constants there. I don't this is expected.

Reply:
I thought that it would be questionable approach :)

My reasoning is that if we're finalizing config — it owns all the values and we should take that into account. IMHO
```
irb(main):005:0> App.config.database.dsn
=> "sqlite:memory"
irb(main):006:0> App.config.database.dsn << 'qwe'
=> "sqlite:memoryqwe"
```
is quite unexpected after we `#finalize!`, especially since documentation says `# Note that finalized configs cannot be changed.` & changelog says `Introduce `Configurable.finalize!` which freezes config and its dependencies`.

Reply:
Yeah, I understand the reasoning, I would be happy to have everything immutable in the first place. But my position on this it's not up to libraries to mutate objects provided by the user. Freezing is a type of mutation so this rule applies. There are two problems: 1) it's unexpected 2) you can't opt-out of this behavior. I don't think we should be unnecessarily smart about this. We could have a warning shown when a non-frozen value passed to the config but then I guess there'd be a ton of them in a typical project, I'm not sure if this would be helpful.

Reply:
Right now semantics of `#finalize!` is not clearly defined, so what is the expected behavior? Freeze just config structure, so we're protected from deleting of used keys?
From the changelog entry I expected that `dependencies` are config values, not sure what was meant by `dependencies` if they're not.
Implicit mutation is certainly not a good thing, but `#finalize!` is explicit user's decision and if semantic would be defined as `freeze all the keys & values` — why not? If you want to opt-out — just don't call `#finalize!`.

If we still need to freeze the structure without values — can we change `#freeze` to freeze the config without values recursively, so the structure is set in stone, and make `#finalize!` to freeze also the values?

Right now `#finalize!` seems broken anyway, because freezing only top-level keys has doubtful usefullness.

Reply:
It's not that simple. `finalized!` is called by dry-system and it's an essinatial part in the lifecycle, you can't skip it. When people install new version of dry-system witth current change they'll get new semantics for the "why not?" reason. I agree it shouldn't break things but the problem is we can't be sure and our semantics will be different from what's understood as a default in ruby: calling `.freeze` on arrays and hashes doesn't freeze values. 

WRT recursive `finalize!` and freezing config structure I don't see problems with this. I agree it's not expected config structure can be changed after freezing.

Reply:
Re dry-system, I believe it no longer depends on this hook (in master) but yeah, if the behavior changes here it'll break current dry-system releases.

Reply:
Yeah, I checked, it's not called. It's not clear whether it should be though. All in all, I think we could have freezing values as an option `.finalize!(freeze_values: true)`. We can debate if it should be `true` or `false` by default but either way, it provides an opt-in/opt-out mechanism which is good. WDYT?

Reply:
You're right that dry-system doesn't call `finalize!` on its config at the moment. This would be worth investigating before we ship 1.0. I'll file an issue.

In fact, I'm not sure if we really finalize configs much at all across the dry-rb gems at the moment.

Having a `freeze_values` to opt in or out of freezing non-`Config` values like a sensible way to go here 👍🏼

Reply:
Issue for dry-system here: https://github.com/dry-rb/dry-system/issues/168